src/main/java/appeng/core/sync/packets/PacketIfaceTermUpdate.java

```diff
@@ -0,0 +1,446 @@
+
+        if (gs instanceof GuiInterfaceTerminal) {
+            ((GuiInterfaceTerminal) gs).postUpdate(this.commands, this.statusFlags);
+        }
+    }
+
+    enum PacketType {
+
+        ADD,
+        REMOVE,
+        OVERWRITE,
+        RENAME;
+
+        public static final PacketType[] TYPES = PacketType.values();
+
+        /**
+         * Get the indexed value of packet type, throws on invalid index
+         */
+        public static PacketType valueOf(int idx) throws ArrayIndexOutOfBoundsException {
+            return TYPES[idx];
+        }
+    }
+
+    /**
+     * A packet for updating an entry.
+     */
+    public abstract static class PacketEntry {
+
+        public final long entryId;
+
+        protected PacketEntry(long entryId) {
+            this.entryId = entryId;
+        }
+
+        protected PacketEntry(ByteBuf buf) throws IOException {
+            this.entryId = buf.readLong();
+            read(buf);
+        }
+
+        /**
+         * Needs to write the packet id.
+         */
+        protected abstract void write(ByteBuf buf) throws IOException;
+
+        /**
+         * Reading the entry id is not needed.
+         */
+        protected abstract void read(ByteBuf buf) throws IOException;
+    }
+
+    /**
+     * A command for sending a new entry.
+     */
+    public static class PacketAdd extends PacketEntry {
+
+        public String name;
+        public int x, y, z, dim;
+        public int rows, rowSize;
+        public boolean online;
+        public ItemStack selfRep, dispRep;
+        public NBTTagList items;
+
+        PacketAdd(long id, String name, boolean online) {
+            super(id);
+            this.name = name;
+            this.online = online;
+        }
+
+        PacketAdd(ByteBuf buf) throws IOException {
+            super(buf);
+        }
+
+        public PacketAdd setLoc(int x, int y, int z, int dim) {
+            this.x = x;
+            this.y = y;
+            this.z = z;
+            this.dim = dim;
+            return this;
+        }
+
+        public PacketAdd setItems(int rows, int rowSize, NBTTagList items) {
+            this.rows = rows;
+            this.rowSize = rowSize;
+            this.items = items;
+
+            return this;
+        }
+
+        public PacketAdd setReps(ItemStack selfRep, ItemStack dispRep) {
+            this.selfRep = selfRep;
+            this.dispRep = dispRep;
+
+            return this;
+        }
+
+        @Override
+        protected void write(ByteBuf buf) throws IOException {
+            buf.writeByte(PacketType.ADD.ordinal());
+            buf.writeLong(entryId);
+            ByteBufUtils.writeUTF8String(buf, this.name);
+            buf.writeInt(x);
+            buf.writeInt(y);
+            buf.writeInt(z);
+            buf.writeInt(dim);
+            buf.writeInt(rows);
+            buf.writeInt(rowSize);
+
+            ByteBuf tempBuf = Unpooled.directBuffer(256);
```

Comment:
I hadn't where this buffer gracefully released. Could you check it for potential leakage

Reply:
I don't remember why this is direct in the first place, should be a heap buffer

Reply:
Ok better to wrap this in try block and gracefully release buffer in finally section

Reply:
heap buffer should be collected by GC already

Reply:
Yes it can be collected by GC but here you put resource allocation on it and buffer should be released it means ref count should be 0 before it can be correctly pass GC phase

Reply:
I don't think its needed but sure, I added a release
See https://github.com/netty/netty/issues/9991